rmiyoyo/catalogue-of-my-things
# Complete UML Implementation (Feature/lily-complete-uml-implementation)

## Complete UML Implementation (Feature/lily-complete-uml-implementation)

### Summary of Changes:

- Created MusicAlbum class in a separate `.rb` file.
- Created Genre class with an association to the Item class (in a separate `.rb` file).
- Defined and set up all MusicAlbum class properties visible in the diagram in the constructor method.
- Defined and set up all Genre class properties visible in the diagram in the constructor method.
- Implemented methods:
  - `add_item` method in the Genre class:
    - Takes an instance of the Item class as input.
    - Adds the input item to the collection of items.
    - Adds self as a property of the item object (using the correct setter from the item object).
  - `can_be_archived?()` in the MusicAlbum class:
    - Overrides the method from the parent class.
    - Returns true if the parent's method returns true AND if `on_spotify` equals true; otherwise, returns false.
- Added unit tests for all implemented methods.
- Provided the following options:
  - List all music albums.
  - List all genres (e.g., 'Comedy', 'Thriller').
  - Add a music album.
  - Add a genre.
- Preserved all data by saving collections in `.json` files.

### Actions Taken:

- Created MusicAlbum class.
- Implemented Genre class.
- Added constructor methods and set up properties.
- Implemented `add_item` and `can_be_archived?` methods.
- Added unit tests for implemented methods.
- Organized the code structure.
- Implemented data preservation through saving collections in `.json` files.

### Testing:

- Ran unit tests for all implemented methods.
- Verified the functionality of options: list all music albums, list all genres, add a music album, add a genre.

### Associated Issues:

Closes #20 #21 #22  #23 #24  #33

### Checklist:

- [x] Unit tests for all implemented methods.
- [x] Data preservation through saving collections in `.json` files.
- [x] Verified functionality of options: list all music albums, list all genres, add a music album, add a genre.

## Review Notes:

### Steps to Run Tests:

1. Ensure all required dependencies are installed.
2. Run the following command to execute unit tests:
   ```bash
   rspec
   ```
    This will run the test suite and provide information on test results.

### Steps to Run the App in the Console:

1. Open a terminal.
2. Navigate to the project directory.
3. Run the main application file using the following command:
   ```bash
   ruby main.rb
   ```
4. Follow the on-screen prompts to interact with the application.

## Additional Notes:
- Confirm that the app preserves data by checking the relevant .json files.
- Pay attention to the console output for any error messages or unexpected behavior.

**_Please review this pull request at your earliest convenience. Your feedback is greatly appreciated.
Thank you for your time and consideration._**
